### PR TITLE
Removing the stream from MessageContext and ErrorContext

### PR DESCRIPTION
**Pull request by danielmarbach**
@bording brought up a good point that for some transports the transport gets a byte array, turns that into a memory stream to pass it into the core just so that incoming message turns that again into a byte array. That doesn't make sense. Let's face it, core is not ready to properly handle streams. This PR backtracks that. WIP PRs all submitted to transports to show the influence.

* Azure Service Bus influence https://github.com/Particular/NServiceBus.AzureServiceBus/pull/285
* Azure Storage Queue influence https://github.com/Particular/NServiceBus.AzureStorageQueues/pull/137
* Sql Server influence https://github.com/Particular/NServiceBus.SqlServer/pull/282
* RabbitMQ influence https://github.com/Particular/NServiceBus.RabbitMQ/pull/186

### CODE REVIEW

**Comment by danielmarbach** [comment]
@Particular/azure-service-bus-maintainers @Particular/azure-storage-queues-maintainers @Particular/rabbitmq-transport-maintainers @Particular/sqlserver-transport-maintainers Thoughts on that one?

**Comment by Scooletz** [comment]
I've seen your PR. For ASQ it's totally ok to go with a byte array.

**Comment by adamralph** [comment]
LGTM from a Rabbit POV.

**Comment by SeanFeldman** [comment]
:+1: for ASB (when updated core is there)

**Comment by tmasternak** [comment]
:+1: from Sql

**Comment by danielmarbach** [comment]
@Particular/azure-service-bus-maintainers @Particular/azure-storage-queues-maintainers @Particular/rabbitmq-transport-maintainers @Particular/sqlserver-transport-maintainers pushed an update. This is now ready to go. One benefit of this approach is that the ErrorContext no longer assumes the stream can be rewind. It is now in the responsibility of the transport to pass in the original body into the error context.

**Comment by adamralph** [comment]
@danielmarbach what was the update? Would it affect https://github.com/Particular/NServiceBus.RabbitMQ/pull/186 in any way?

**Comment by danielmarbach** [comment]
@adamralph I made it work for MSMQ :dancer:  and it is no longer WIP ;)

**Comment by adamralph** [comment]
@danielmarbach if that means no change for https://github.com/Particular/NServiceBus.RabbitMQ/pull/186 then I'm :+1: from a RabbitMQ transport perspective.

**Comment by danielmarbach** [comment]
Well @adamralph this would enable you to do `byte[]` buffering :trollface: in Rabbit

File: src/NServiceBus.Core.Tests/DelayedDelivery/TimeoutManager/DispatchTimeoutBehaviorTest.cs [hunk 1]
```diff
@@ -135,19 +134,22 @@ public Task Dispatch(TransportOperations outgoingMessages, ContextBag context)
...
+            public Task Dispatch(IEnumerable<TransportOperation> outgoingMessages, ContextBag context)
             {
                 throw new Exception("simulated exception");
             }
         }
 
         class FakeTimeoutStorage : IPersistTimeouts
         {
+            public Func<string, ContextBag, bool> OnTryRemove { get; set; } = (id, bag) => true;
+            public Func<string, ContextBag, TimeoutData> OnPeek { get; set; } = (id, bag) => null;
```
Comment (timbussmann) [inline src/NServiceBus.Core.Tests/DelayedDelivery/TimeoutManager/DispatchTimeoutBehaviorTest.cs]: why are these moved up?

Comment (danielmarbach) [inline src/NServiceBus.Core.Tests/DelayedDelivery/TimeoutManager/DispatchTimeoutBehaviorTest.cs]: I think I did a cleanup. Will check

File: src/NServiceBus.Core/DelayedDelivery/TimeoutManager/StoreTimeoutBehavior.cs [hunk 2]
```diff
@@ -50,15 +50,12 @@ public StoreTimeoutBehavior(ExpiredTimeoutsPoller poller, IDispatchMessages disp
...
-                var body = new byte[context.BodyStream.Length];
-
-                await context.BodyStream.ReadAsync(body, 0, body.Length).ConfigureAwait(false);
 
                 var data = new TimeoutData
                 {
                     Destination = destination,
                     SagaId = sagaId,
-                    State = body,
+                    State = context.Body, // TODO: Need for copying?
```
Comment (timbussmann) [inline src/NServiceBus.Core/DelayedDelivery/TimeoutManager/StoreTimeoutBehavior.cs]: remove the todo and maybe raise an issue about this?

File: src/NServiceBus.Core/Recoverability/Faults/ErrorsNotifications.cs [hunk 3]
```diff
@@ -55,6 +55,7 @@ internal void InvokeMessageHasBeenSentToSecondLevelRetries(int secondLevelRetryA
             MessageHasBeenSentToSecondLevelRetries?.Invoke(this, secondLevelRetry);
         }
 
+        // TODO: Really necessary/
```
Comment (timbussmann) [inline src/NServiceBus.Core/Recoverability/Faults/ErrorsNotifications.cs]: ?

File: src/NServiceBus.Core/DelayedDelivery/TimeoutManager/StoreTimeoutBehavior.cs (hunk 2 above)
Comment (danielmarbach) [inline src/NServiceBus.Core/DelayedDelivery/TimeoutManager/StoreTimeoutBehavior.cs]: :+1:

**Comment by timbussmann** [comment]
Looks good to me besides some minor comments. :+1:

File: src/NServiceBus.Core.Tests/DelayedDelivery/TimeoutManager/DispatchTimeoutBehaviorTest.cs (hunk 1 above)
Comment (danielmarbach) [inline src/NServiceBus.Core.Tests/DelayedDelivery/TimeoutManager/DispatchTimeoutBehaviorTest.cs]: Ran cleanup again and it remained like that

**Comment by danielmarbach** [comment]
@timbussmann updated. Will raise an issue

File: src/NServiceBus.Core.Tests/DelayedDelivery/TimeoutManager/DispatchTimeoutBehaviorTest.cs (hunk 1 above)
Comment (timbussmann) [inline src/NServiceBus.Core.Tests/DelayedDelivery/TimeoutManager/DispatchTimeoutBehaviorTest.cs]: ah, because these are properties -.- all good.

**Comment by adamralph** [comment]
I'll update https://github.com/Particular/NServiceBus.RabbitMQ/pull/181
